torch/utils/checkpoint.py

```diff
@@ -151,19 +151,25 @@ def checkpoint(function, *args, use_reentrant: bool = True, **kwargs):
     and instead recomputes them in backward pass. It can be applied on any part
     of a model.
 
-    Specifically, in the forward pass, :attr:`function` will run in
+    Depending on the flag `use_reentrant`, one of two implementations will be
+    invoked. If `use_reentrant=True`, in the forward pass, :attr:`function` will run in
     :func:`torch.no_grad` manner, i.e., not storing the intermediate
     activations. Instead, the forward pass saves the inputs tuple and the
     :attr:`function` parameter. In the backwards pass, the saved inputs and
     :attr:`function` is retrieved, and the forward pass is computed on
     :attr:`function` again, now tracking the intermediate activations, and then
     the gradients are calculated using these activation values.
+    If `use_reentrant=False`, an implementation based on
+    :func:`torch.autograd.graph.saved_tensors_hooks` will be used, resulting in
+    references to intermediate activations not being saved in the forward pass
+    and being recomputed in the backwards pass.
 
     The output of :attr:`function` can contain non-Tensor values and gradient
-    recording is only performed for the Tensor values. Note that if the output
-    consists of nested structures (ex: custom objects, lists, dicts etc.)
-    consisting of Tensors, these Tensors nested in custom structures will not
-    be considered as part of autograd.
+    recording is only performed for the Tensor values. Note that if
+    ``use_reentrant=True`` and the output consists of nested structures
```

Comment:
nit I think this could be slightly simplified.
Note that if ``use_reentrant=True` only Tensors that are passed directly as input/output are considered and anything inside data structures (list, tuple, dict, etc) will be ignored.